### ISSUE TITLE
[clang-tidy] False positive bugprone-exception-escape for throwing lambda in noexcept function

### ISSUE THREAD

**Issue by Eisenwave** [llvm/llvm-project#132605]
https://godbolt.org/z/1sf8Gne35
```cpp
void awoo() noexcept {
    [] { throw 0; };
}
```
This outputs
```
<source>:1:6: warning: an exception may be thrown in function 'awoo' which should not throw exceptions [bugprone-exception-escape]
    1 | void awoo() noexcept {
      |      ^
```

**Comment by denzor200** [comment]
I tried to reproduce the issue in a simplified clang-query request. Does anybody can explain why the simple matcher `functionDecl(isNoThrow())` triggers on lambda like `[]() noexcept(false) { throw 0; };` ? While it doesn't trigger on a regular function with `noexcept(false)`

https://godbolt.org/z/j3ac791qo

Is it a bug on the AST side?

**Comment by vbvictor** [comment]
> I tried to reproduce the issue in a simplified clang-query request. Does anybody can explain why the simple matcher `functionDecl(isNoThrow())` triggers on lambda like `[]() noexcept(false) { throw 0; };` ? While it doesn't trigger on a regular function with `noexcept(false)`

In your example, only `CXXConversionDecl` and `CXXDestructorDecl` are matched as `noexcept` - https://godbolt.org/z/aefG99dYv. The main `operator()` with `throw` expression is not matched.

**Comment by vbvictor** [comment]
Note: code example in issue is only false-positive when lambda is not assigned to a variable.
```cpp
void awoo() noexcept { // no warning
    auto l = [] { throw 0; };
}

void awoo() noexcept { // warning
    auto l = [] { throw 0; };
    l();
}

void awoo() noexcept { // warning
    [] { throw 0; } ();
}
```
